rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -396,14 +418,14 @@ class rmw_context_impl_s::Data final
   // Enclave, name used to find security artifacts in a sros2 keystore.
   std::string enclave_;
   // An owned session.
-  z_owned_session_t session_;
+  std::shared_ptr<zenoh::Session> session_;
   // An optional SHM manager that is initialized of SHM is enabled in the
   // zenoh session config.
-  std::optional<z_owned_shm_provider_t> shm_provider_;
+  std::optional<zenoh::ShmProvider> shm_provider_;
   // Graph cache.
   std::shared_ptr<rmw_zenoh_cpp::GraphCache> graph_cache_;
   // ROS graph liveliness subscriber.
-  z_owned_subscriber_t graph_subscriber_;
+  std::optional<zenoh::Subscriber<void>> graph_subscriber_cpp_;
```

Comment:
This should not be an optional. `make()` should fail if we are unable to construct the sub.

Reply:
There is no public contructor for this class

Reply:
I share @Yadunund 's uncomfortable-ness with this type.  It doesn't seem correct.

However, this won't actually compile if this is just a `zenoh::Subscriber<void>`, because then the constructor for the `Data` class will try to call the empty constructor on `Subscriber`, which doesn't exist.  By wrapping this in an `std::optional`, we workaround this by basically calling the empty constructor for `std::optional` during `Data` construction, and then filling it in for real later on.

The only other way to do this that is coming to mind is to make a new version of `liveliness_declare_subscriber` which returns a pointer, and then to make this a `std::shared_ptr` or something.  But that isn't intrinsically better in my mind.

To fix this for real we'd probably have to restructure the class hierarchy here so that we had one class that contained the `Session`, and another that contained the `Subscriber`.  But I think that is too much to do in this PR, so the only thing I'll request is a comment here explaining why we are doing things this way:

```suggestion
  // The graph_subscriber *must* exist in order for anything in this Data class, and hence rmw_zenoh_cpp, to work.
  // However, zenoh::Subscriber does not have an empty constructor, so just declaring this as a zenoh::Subscriber fails to compile.
  // We work around that by wrapping it in a std::optional, so the std::optional gets constructed at Data constructor time,
  // and then we initialize graph_subscriber_cpp_ later.  Note that the zenoh-cpp API liveliness_declare_subscriber() throws an
  // exception if it fails, so this should all be safe to do.
  std::optional<zenoh::Subscriber<void>> graph_subscriber_cpp_;
```